e2e/libs/keywords/statefulset_keywords.py

```diff
@@ -0,0 +1,141 @@
+from workload.persistentvolumeclaim import delete_persistentvolumeclaim
+from workload.statefulset import create_statefulset
+from workload.statefulset import delete_statefulset
+from workload.statefulset import get_statefulset
+from workload.statefulset import scale_statefulset
+from workload.statefulset import wait_for_statefulset_replicas_ready
+from workload.workload import check_pod_data_checksum
+from workload.workload import get_workload_pod_names
+from workload.workload import get_workload_pvc_name
+from workload.workload import get_workload_volume_name
+from workload.workload import keep_writing_pod_data
+from workload.workload import write_pod_random_data
+
+from persistentvolumeclaim.persistentvolumeclaim import PersistentVolumeClaim
+
+from volume.volume import Volume
+
+from volume.constant import MEBIBYTE
+
+from utility.resource_cache import ResourceCache
+
+from utility.constant import CACHE_CHECKSUM
+from utility.constant import CACHE_EXPANDED_SIZE
+from utility.constant import CACHE_ORIGINAL_REPLICA_COUNT
+from utility.constant import KIND_PERSISTENTVOLUMECLAIM
+from utility.constant import KIND_POD
+from utility.constant import KIND_STATEFULSET
+from utility.constant import KIND_VOLUME
+
+
+class statefulset_keywords:
+
+    def __init__(self):
+        self.volume = Volume()
+        self.persistentvolumeclaim = PersistentVolumeClaim()
+
+    def cleanup_statefulsets(self):
+        cached = ResourceCache.get_kind(KIND_STATEFULSET)
+        if cached is None:
+            return
+
+        uncache = []
+        for name, links in cached.items():
+            delete_statefulset(name)
+            uncache.append(name)
+
+            if KIND_PERSISTENTVOLUMECLAIM in links:
+                pvc_name = links[KIND_PERSISTENTVOLUMECLAIM]
+                delete_persistentvolumeclaim(pvc_name)
+                ResourceCache.delete(KIND_PERSISTENTVOLUMECLAIM, pvc_name)
+
+        for name in uncache:
+            ResourceCache.delete(KIND_STATEFULSET, name)
+        assert len(ResourceCache.get_kind(KIND_STATEFULSET)) == 0
+
+    def create_statefulset(self, name, volume_type="rwo", option=""):
+        create_statefulset(name, volume_type, option)
+        ResourceCache.new(KIND_STATEFULSET, name)
+
+        pod_name = get_workload_pod_names(name)[0]
+        ResourceCache.new(KIND_POD, pod_name)
+        ResourceCache.create_or_update_link(
+            KIND_STATEFULSET, name, KIND_POD, pod_name
+        )
+
+        volume_name = get_workload_volume_name(name)
+        ResourceCache.new(KIND_VOLUME, volume_name)
+        ResourceCache.create_or_update_link(
+            KIND_STATEFULSET, name, KIND_VOLUME, volume_name
+        )
+
+        pvc_name = get_workload_pvc_name(name)
+        ResourceCache.new(KIND_PERSISTENTVOLUMECLAIM, pvc_name)
```

Comment:
Sure, however, the current implementation `get_workload_pvc_name` doesn't cover the case when the statefulset is scaled down. Considering that the PVC name typically remains static for the workload, it might be for efficient to utilize the cache instead of repeatedly querying the API server for various resources just to retrieve the PVC name.

Reply:
Can you help to elaborate more on what's the drawback of repeatedly querying the API server? Isn't it the same as calling kubectl in our daily operations? Thanks!